core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/Envelope.kt

```diff
@@ -0,0 +1,414 @@
+                    beginSpeed,
+                    endStepIndex,
+                    endPosition,
+                    endSpeed
+                ) ?: return arrayOf()
+            return arrayOf(sliced)
+        }
+
+        val beginPart = parts[beginPartIndex]
+        val endPart = parts[endPartIndex]
+        val beginPartSliced = beginPart.sliceEnd(beginStepIndex, beginPosition, beginSpeed)
+        val endPartSliced = endPart.sliceBeginning(endStepIndex, endPosition, endSpeed)
+
+        // compute the number of unchanged envelope parts between sliced parts
+        val copySize = endPartIndex - beginPartIndex + 1 - /* sliced endpoints */ 2
+
+        val res = arrayListOf<EnvelopePart>()
+        if (beginPartSliced != null) res.add(beginPartSliced)
+        val copyStartIndex = beginPartIndex + 1
+        for (i in 0 until copySize) res.add(parts[copyStartIndex + i])
+        if (endPartSliced != null) res.add(endPartSliced)
+
+        return res.toTypedArray()
+    }
+
+    // endregion
+    override fun iterator(): MutableIterator<EnvelopePart> {
+        return object : MutableIterator<EnvelopePart> {
+            private var i = 0
+
+            override fun hasNext(): Boolean {
+                return i < parts.size
+            }
+
+            override fun next(): EnvelopePart {
+                if (!hasNext()) throw NoSuchElementException()
+                return parts[i++]
+            }
+
+            override fun remove() {
+                TODO("Not yet implemented")
+            }
+        }
+    }
+
+    override fun iteratePoints(): List<EnvelopeTimeInterpolate.EnvelopePoint> {
+        val res = ArrayList<EnvelopeTimeInterpolate.EnvelopePoint>()
+        var time = 0.0
+        for (part in this) {
+            // Add head position points
+            for (i in 0 until part.pointCount()) {
+                val pos = part.getPointPos(i)
+                val speed = part.getPointSpeed(i)
+                res.add(EnvelopeTimeInterpolate.EnvelopePoint(time, speed, pos))
+                if (i < part.stepCount()) time += part.getStepTime(i)
+            }
+        }
+        return res
+    }
+
+    /** Makes a stream from the parts */
+    fun stream(): Stream<EnvelopePart> {
+        return Arrays.stream(parts)
+    }
+
+    companion object {
+        /** Create a new Envelope */
+        @JvmStatic
+        fun make(vararg parts: EnvelopePart): Envelope {
+            return Envelope(arrayOf(*parts))
+        }
+    }
+}
+
+/** Build the min envelopes between 2 envelopes starting and ending at the exact same positions. */
+fun minEnvelopes(envelopeA: Envelope, envelopeB: Envelope): Envelope {
+    val minParts = mutableListOf<EnvelopePart>()
+    assert(envelopeA.beginPos == envelopeB.beginPos)
+    assert(envelopeA.endPos == envelopeB.endPos)
+
+    val keyPositions =
+        Arrays.stream(envelopeA.partPositions)
+            .boxed()
+            .collect(Collectors.toCollection { TreeSet() })
```

Comment:
```suggestion
    val keyPositions = TreeSet(envelopeA.partPositions.toList())
```

It does have a small performance overhead compared to the stream (hoping that the stream version avoids copies), it's just a lot less clutter. Otherwise we could create a small utility function to hide it away.

(same for the line just below)